### PR TITLE
feat(dashboard): implement widget item api v2

### PR DESCRIPTION
**Pull request by st3iny**
* Resolves: _none_

## Summary

This API enables the dashboard to render all widgets from the API data alone without having apps to provide their own bundles. This saves a lot of traffic and execution time as a lot less javascript has to be parsed on the frontend.

This is already done on the clients. However, the dashboard requires some extra logic and data as its widgets tend to be more sophisticated. To prevent breaking all apps and clients new API (v2) was introduced allowing each app and each client to migrate at their own pace.

## Profiling results

results

**Perceived loading time:** Time from hitting refresh until **all** dashboard widgets finished loading.
**No cache:** Simulates a new visitor or a visitor returning after a long time. Cache is disabled in Chrome DevTools to force the browser to redownload all assets.
**Cache:** Simulates a user that often uses Nextcloud. Browser cache is enabled.

Tests were run on the same instance with exactly the same calendar, mail and spreed data etc.

### Without cache

**Time until all widgets finised loading:** 4284 ms

**Chrome network stats:**
```
151 requests
36.8 MB transferred
36.8 MB resources
DOMContentLoaded: 2.09 s
Load: 3.72 s
```

### With cache

**Time until all widgets finised loading:** 3836 ms

**Chrome network stats:**
```
190 requests
126 kB transferred
36.6 MB resources
DOMContentLoaded: 1.34 s
Load: 2.76 s
```

### Without cache

**Time until all widgets finished loading:** 2995 ms

**Chrome network stats:**
```
111 requests
29.0 MB transferred
28.9 MB resources
DOMContentLoaded: 1.63 s
Load: 3.05 s
```

### With cache

**Time until all widgets finished loading:** 1500 ms

**Chrome network stats:**
```
132 requests
75.8 kB transferred
28.9 MB resources
DOMContentLoaded: 919 ms
Load: 1.53 s
```

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/public/Dashboard/IReloadableWidget.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+namespace OCP\Dashboard;
+
+/**
+ * Allow {@see IAPIWidgetV2} to reload their items
+ *
+ * @since 27.1.0
+ */
+interface IReloadableWidget extends IWidget {
```
Comment (tcitworld) [inline lib/public/Dashboard/IReloadableWidget.php:34]: **question:** What's the advantage of having it's own interface? Implementing `IReloadableWidget` but not `IAPIWidgetV2`?

Comment (st3iny) [reply lib/public/Dashboard/IReloadableWidget.php:34]: That is a good point. Moving the reload interval logic inside `IAPIWidgetV2` makes sense and spares an interface.

Done.

File: lib/composer/composer/autoload_psr4.php [hunk 2]
```diff
@@ -9,5 +9,6 @@
     'OC\\Core\\' => array($baseDir . '/core'),
     'OC\\' => array($baseDir . '/lib/private'),
     'OCP\\' => array($baseDir . '/lib/public'),
+    'Bamarni\\Composer\\Bin\\' => array($vendorDir . '/bamarni/composer-bin-plugin/src'),
```
Comment (nickvergessen) [inline lib/composer/composer/autoload_psr4.php:12]: `composer install --no-dev`

Comment (st3iny) [reply lib/composer/composer/autoload_psr4.php:12]: Done

**Review by provokateurin — APPROVED** [review]
API changes LGTM

File: lib/public/Dashboard/IReloadableWidget.php (hunk 1 above)
Comment (st3iny) [reply lib/public/Dashboard/IReloadableWidget.php:34]: To have it as an opt in feature. It doesn't make sense for all widgets to reload. I changed the interface to extend `IAPIWidgetV2` to make it clear that only migrated widgets can make use of this interface.

File: lib/composer/composer/autoload_psr4.php (hunk 2 above)
Comment (nickvergessen) [reply lib/composer/composer/autoload_psr4.php:12]: Hmm still here, maybe it was forgotten on another PR

**Review by nickvergessen — APPROVED** [review]
Needs a rebase, but otherwise seems fine and functional

**Review by AndyScherzinger — APPROVED** [review]

**Comment by nickvergessen** [comment]
```
Conflicting files
dist/settings-users-8351.js
dist/settings-users-8351.js.map
dist/settings-vue-settings-apps-users-management.js
dist/settings-vue-settings-apps-users-management.js.map 
```

Rebasing

**Comment by AndyScherzinger** [comment]
/backport 6982597b6a6d319dacfbe3bee2edd2a39b3d6d68 to stable27

**Comment by nickvergessen** [comment]
Will have to do it manually, the OpenAPI stuff does not exist on stable27

**Comment by nickvergessen** [comment]
PS I'm on it aklready

**Pull request merged**